doc/dev/static_type_checking.md

```diff
@@ -0,0 +1,1399 @@
+   what we say we return - it's more helpful to specify concrete types which a user can reason about than abstract types.
+
+2) It can be more useful to consider the set of supported operations as the defining characteristic of a type.
+   [Structural subtyping](https://en.wikipedia.org/wiki/Structural_type_system) is what supports duck typing in
+   Python.
+
+#### Mapping types
+
+Syntax for mapping types follows: `MappingType[Key, Value]`. For example, `typing.Dict[str, str]` indicates that
+the type is a dictionary with string keys and string values.
+
+For mapping parameter types, consider using `Mapping` or `MutableMapping`. If the function accepting the mapping
+type does not need to mutate the mapping, `Mapping` provides more flexibility than `MutableMapping` (which
+implements methods like `pop`, `update`, `clear`).
+
+For example,
+
+```python
+from typing import Mapping
+
+def create_table(entity: Mapping[str, str]) -> None:
+    ...
+```
+
+By accepting `Mapping` here, we give more flexibility to the user and allow them to specify anything consistent
+with a `Mapping` - dict, defaultdict, a user-defined dict-like class, or any subtype of `Mapping`. Had we
+specified a generic `typing.Dict` here, the `entity` passed must be a `dict` or one of its subtypes, narrowing the types
+accepted.
+
+With [PEP 589](https://peps.python.org/pep-0589/), a `TypedDict` was introduced in Python 3.8 which allows you to add
+type hints for dictionaries with a fixed set of keys. The syntax for this looks similar to building a dataclass:
+
+```python
+# from typing import TypedDict  # Python >= 3.8
+from typing_extensions import TypedDict
+
+
+class Employee(TypedDict):
+    name: str
+    title: str
+    id: int
+    current: bool
+```
+
+A `TypedDict` lets you specify types hints for the value of each key in the dictionary.
+It's important to note that `Employee` has no runtime effect - the interpreter will not validate that the keys or values
+passed are the correct types. However, during static type checking, when the `Employee` TypedDict is constructed, the type checker
+will expect the keys and typed values specified.
+
+```python
+employee = Employee(name="krista", title="swe", id="nah", current=True)
+```
+
+mypy identifies that the `id` key is the wrong type:
+
+`main.py:9: error: Incompatible types (expression has type "str", TypedDict item "id" has type "int")`
+
+At runtime, `employee` is perfectly valid even though it does not conform to a true `Employee`:
+
+```cmd
+>>>employee = Employee(name="krista", title="swe", ident="nah", current=True)
+>>>print(employee)
+{'name': 'krista', 'title': 'swe', 'ident': 'nah', 'current': True}
+>>>type(employee)
+<class 'dict'>
+```
+
+Note that this creates a plain `dict` at runtime. Usage of `TypedDict` is a great way to inform the type checkers and
+Intellisense how a specific dict should be constructed and help the type checkers warn users if they try to access
+keys which don't exist.
```

Comment:
It might be worth explicitly calling out here that because this is a plain `dict` at runtime, you cannot do an `isinstance` check, e.g. `isinstance(employee, Employee)`